flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/source/assigner/EventTimeAlignedAssigner.java

```diff
@@ -0,0 +1,301 @@
+/*
+ *
+ *  * Licensed to the Apache Software Foundation (ASF) under one
+ *  * or more contributor license agreements.  See the NOTICE file
+ *  * distributed with this work for additional information
+ *  * regarding copyright ownership.  The ASF licenses this file
+ *  * to you under the Apache License, Version 2.0 (the
+ *  * "License"); you may not use this file except in compliance
+ *  * with the License.  You may obtain a copy of the License at
+ *  *
+ *  *   http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing,
+ *  * software distributed under the License is distributed on an
+ *  * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  * KIND, either express or implied.  See the License for the
+ *  * specific language governing permissions and limitations
+ *  * under the License.
+ *
+ */
+
+package org.apache.iceberg.flink.source.assigner;
+
+import java.io.Serializable;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.apache.flink.api.common.eventtime.TimestampAssigner;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplitState;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplitStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.flink.calcite.shaded.com.google.common.collect.Iterables.getFirst;
+
+public class EventTimeAlignedAssigner implements SplitAssigner, WatermarkTracker.Listener {
```

Comment:
I also thinking if we should call this just `TimeAlignedAssigner` or `WatermarkAlignedAssigner`. I know some use cases can use the snapshot timestamp as the order time. It doesn't have to be the min-max column stats.

what do you think? @pvary

Reply:
I prefer `WatermarkAlignedAssigner` too.

But TBH, I am still not sure how we will extract the information from the splits, and it is more concerning for me. Will check the internal implementation today, to see if it helps 😀

Reply:
the only problem with `Watermark` is that this watermark is internal to the source/assigner. it is not the actual Flink watermark. might cause confusion to users.

The time/watermark is extracted from the min value of the file/split using column min-max stats. I also replied on how to handle combined file scan task. essentially, we like to avoid multiple files in one scan task.

The time extraction can also be the snapshot timestamp that the file was appended. I believe snapshot timestamp and sequence number are not exposed today, which is sth we need to change in the Iceberg core for scan planning.

Reply:
Yes, as a user, I originally thought the use of "Watermark" in this code meant it was the same as a Flink Watermark. It took a while to understand they aren't the same, although I understand why the name choice makes sense.

Reply:
What if we generate real Flink Watermarks based on the Split statistics and emit them alongside the records at every split start:
```
  @Override
  public void emitRecord(
      RecordAndPosition<T> element, SourceOutput<T> output, IcebergSourceSplit split) {
    if (timestampAssigner != null && !split.splitId().equals(lastSplit)) {
      output.emitWatermark(new Watermark(timestampAssigner.extractTimestamp(split, -1)));
      lastSplit = split.splitId();
    }

    output.collect(element.record());
    split.updatePosition(element.fileOffset(), element.recordOffset());
  }
```

Then we can use [Flink Watermark Alignment](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/datastream/event-time/generating_watermarks/#watermark-alignment-_beta_) to throttle readers.

With this we can have a very limited Iceberg specific code and rely on Flink for tracking the status of the readers.

CC: @gyfora

Reply:
@pvary It would be great if we could generate real Flink watermarks within the source.

I believe one complication, though, is that rather than emitting the `extractTimestamp` value for each new `splitId`, the emitted watermarks need to (somehow) follow the same logic as `WatermarkTracker`'s `updateWatermark` method. When a split is completed, that split is removed from `assignedSplits` by `onCompletedSplits`. The minimum of all minimum timestamps of assigned + unassigned splits gives a value that correctly delineates all completed data from all possible future data (a proper watermark), and this is what `updateWatermark` computes.

Only the _maximum_ timestamp of a potential split `S_i` is constrained by the split assignment constraint `max_timestamp(S_i) - min(min_timestamp({assigned}), min_timestamp({unassigned})) < maxMisalignmentThreshold`. So, the _minimum_ timestamp of `S_i` that `extractTimestamp` yields could be larger than the proper watermark value.

Since only the assigner is aware of all assigned & unassigned splits, and it's instantiated inside the SplitEnumerator, which runs on the JobManager, some way to communicate watermark updates to the SourceReaders running on the TaskManagers could make it work. Then, perhaps they could call `emitWatermark` with the latest global watermark value they've received whenever a new split is assigned.

Reply:
> The minimum of all minimum timestamps of assigned + unassigned splits gives a value that correctly delineates all completed data from all possible future data

I think this is what Flink Watermark aggregation does. Every Operator makes sure that its own watermark is correct in its own scope (increasing monotonously), and Flink aggregates them and keeps track of the lowest value.

With this watermark, the  watermark alignment pauses the reader which are over the alignment limit. In the current Flink implementation the watermark is aggregated and published asynchronously, so the current implementation does not guarantee that we do not read records above the limit, but we do not read too many.